Add realm setting for advertising org in communities directory.

laurynmm opened the pull request:
Adds a realm setting for organizations to opt in to public communities directory: `realm.list_on_public_communities_directory`.

Prep PR for #15844 (creating the public community directory page).

Still working on how best to update the setting that notifies the front end / client of an organization's eligibility because it depends on a number of moving parts:
- being on Zulip Cloud (updating `realm.plan_type` can change this value)
- either having one or more web-public streams (adding or removing a web-public stream or updating `realm.enable_spectator_access` can change this value)
- allowing users to join without an invitation (updating `realm.invite_required` can change this value)

Other tasks \ questions TBD:
- ~Does removing `public` and shortening to `communities_directory` make sense?~ **Shortened to `communities_directory`**.
- Front end does not live update the enabling / disabling of the checkbox without a page refresh (see notes above)
- ~Spacing on the Organization profile admin page could be improved~ **Fine as is**.
- ~API changelog and version.py bump with final updates to the OpenAPI are TBD in first draft~ **Updated for these**.

### WIP front end

Question - There was a capitalization issue with the original label for the checkbox: "List organization on zulip.com" because "Zulip" is considered a proper noun. Does the updated string work? Do we want something else?

Answer - Updated to be "List organization in the Zulip communities directory."

![Screenshot from 2022-04-27 19-50-09](https://user-images.githubusercontent.com/63245456/165588549-e191e9ca-4735-4a57-9d2a-299f76607471.png)

### Help center documentation updates
1. New article about the **Communities directory** (has shared intro text)
3. Added a section with intro text / links to **Moderating open organizations** and **Create your organization profile**

![Screenshot from 2022-04-28 16-34-12](https://user-images.githubusercontent.com/63245456/165782586-13bd062c-759b-4faa-a8ac-38f5c065399d.png)
![Screenshot from 2022-04-28 16-34-45](https://user-images.githubusercontent.com/63245456/165782592-fa9f48aa-2635-4e9a-87b3-56a79bac7849.png)


### API documentation updates
1. Realm update event for `list_on_public_communities_directory`
2. Register queue realm data `realm_list_on_public_communities_directory`

![Screenshot from 2022-04-28 16-50-12](https://user-images.githubusercontent.com/63245456/165782652-a797374a-ce15-4b48-aeb5-e427558d6265.png)
![Screenshot from 2022-04-28 16-50-44](https://user-images.githubusercontent.com/63245456/165782658-dcc2e224-4264-4ea2-aa71-bda8365a4147.png)


**Self-review checklist**

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] Self-reviewed the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/version-control.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.

alya commented:
Cool, thanks! I'll probably do an edit pass on the documentation later today.

timabbott commented on zerver/lib/events.py:
```diff
@@ -298,6 +298,9 @@ def fetch_initial_state_data(
         state["server_web_public_streams_enabled"] = realm.web_public_streams_available_for_realm()
+        state[
+            "realm_eligible_for_public_communities_directory"
+        ] = realm.eligible_for_public_communities_directory()
```
Looking at this, I'm definitely liking the idea of calculating this from other properties on the frontend (and allowing changing the setting here, it even if 0 web public streams are enabled -- it'll just not be effective yet).

I think from a UX perspective, that's what we'll want, because otherwise the sequence of steps to enable this setting is a pain -- first enable web public streams, then go in and make one, then find this setting and toggle it.  @alya to confirm.

laurynmm replied:
Based on [this CZO conversation](https://chat.zulip.org/#narrow/stream/3-backend/topic/communities.20directory.20eligibility/near/1374482), I moved the calculating eligibility to the front end code. Currently, it's implemented based on the `invite_required` and `enable_spectator_access` realm properties.

I did not limit it based on the `plan_type`, but we could also add that, which would mean self-hosted realms would always have it disabled.

This implementation would require Zulip to do the checking for the existence of at least one web-public stream before actually listing an organization in the directory.

timabbott commented on zerver/views/realm.py:
```diff
@@ -143,6 +143,9 @@ def update_realm(
     enable_spectator_access: Optional[bool] = REQ(json_validator=check_bool, default=None),
+    list_on_public_communities_directory: Optional[bool] = REQ(
+        json_validator=check_bool, default=None
+    ),
```
The API docs say this can only be enabled in certain situations; but I think the correct story is that it only does something useful in those situations, and the API docs should instead describe (1) that it only does something in those cases and (2) that settings UI should only offer it in appropriate situations.

laurynmm replied:
Agreed. I've updated the descriptions (and screenshots) of the API documentation for these changes.

timabbott commented on templates/zerver/help/public-communities-directory.md:
```diff
@@ -0,0 +1,36 @@
+!!! warn ""
+
+    Opting in does not guarantee that an organization will be listed
+    on Zulip's communities directory.
```
I like the idea of adjusting this phrase to explain that "it gives Zulip permission to list you" or something.

timabbott commented:
This generally looks good; just posted a few notes above.

laurynmm commented:
@timabbott Thanks for the review / feedback. I've updated for all those changes.

@alya It would be good to get your feedback on allowing users to opt in even if there are no web-public streams to simplify the UX for enabling the setting.

Here's a screen grab of the current implementation:
![Peek 2022-04-28 16-22](https://user-images.githubusercontent.com/63245456/165784758-a0f258b0-838b-4ef3-b79f-d96b99a46a24.gif)

laurynmm commented on zerver/models.py:
```diff
@@ -980,6 +985,14 @@ def allow_web_public_streams_access(self) -> bool:
+            # directory is limited to Zulip Cloud.
+            return False
+
+        return self.has_web_public_streams() or not self.invite_required
```
I'm not sure if this is necessary now, but thought I would check before removing it and the test I wrote.

laurynmm replied:
I have since removed this function.

laurynmm commented:
Yeah, I was AFK when I saw the failure. Just got back and pushed the link correction.

alya commented:
Ah, yeah, gating the *listing* but not the *setting* on whether or not there is at least one web-public stream sounds like a great solution.

alya commented:
OK, playing around with the PR and chatting with @timabbott , I think we actually need to change the UX a bit. At present, it's very confusing what it means for the checkbox to be selected but disabled.

I think the best solution is not to disable the checkbox at all. Instead, we can just explain on the help page that we currently only plan to list orgs that have a web-public stream or open sign-ups. But regardless, you'll be able to check/uncheck it whenever. Having it checked doesn't *guarantee* that you'll be listed regardless, so the contract with the user about what the checkbox means is a bit complicated no matter what.

Separately, let's rename the setting to: "Advertise organization in the Zulip communities directory". It'll make the value prop a bit more clear.

alya commented:
I also noticed that the question mark as aligned too low for some reason. Note that, in contrast, the alignment is fine on the question mark for the web-public streams checkbox.

![Screen Shot 2022-04-28 at 12 39 56 PM](https://user-images.githubusercontent.com/2090066/165836507-3f58dace-3d59-4da8-b1dc-4e48dbcdf02f.png)
![Screen Shot 2022-04-28 at 12 40 08 PM](https://user-images.githubusercontent.com/2090066/165836518-66b1a6f9-20a3-48c5-b871-7e5aa9690fd2.png)

laurynmm commented:
Most recent updates:
- The realm setting / property is `advertise_in_communities_directory` now and language has been adjusted for such. 'Advertise', 'promote' and 'give permission' are the general verbs that I've been using in documentation and comments.
- The checkbox in the UI is always active and saves / updates like other realm properties.
- Help center documentation commit is now the final commit in the series.
  - It has a shared intro description about the directory, which can be updated for when the directory ready to be launched.
  - It currently has a link to the main help center article on the communities directory (where it is also used as the intro) that I intend to link to the directory page once it is live.
  - The shared intro is used in the articles for **Create your organization profile** and **Moderating open organizations**.
  - Updated screenshots of main article and shared intro are below.

Re: alignment of question mark / help link
- I can't find any differences in the CSS settings for these checkboxes / help links. Here's an image of a few of these settings all together. I'm wondering if the parenthesis changes the perception of the alignment?

![Screenshot from 2022-04-29 13-07-11](https://user-images.githubusercontent.com/63245456/165942728-cf6f1a27-3f69-4f5a-95a2-6d1a7e57d433.png)


### Updated help center documentation screenshots:

![Screenshot from 2022-04-29 14-18-08](https://user-images.githubusercontent.com/63245456/165942923-f63f329b-c53d-47aa-b8c1-ee9a1c6be39a.png)
![Screenshot from 2022-04-29 14-18-30](https://user-images.githubusercontent.com/63245456/165942927-418c63d8-d946-471a-850a-a264027bf1cb.png)

alya commented:
> I can't find any differences in the CSS settings for these checkboxes / help links. Here's an image of a few of these settings all together. I'm wondering if the parenthesis changes the perception of the alignment?

Hmm, you might be right! Let's not worry about it.

laurynmm commented:
Once the CI issue is resolved, I'll push the rebased version for the changes to the OpenAPI documentation.

timabbott commented:
I think probably I'm going to proceed with just trying to merge this; the CI failure affecting `main` may take some time to resolve.

timabbott commented:
I addressed the items noted above. I have a few thoughts before merging:
* I think we should squash the commits; the /help/ content is linked to from the others, and there's no real benefit to having multiple commits. (It's just as easy to review the whole block together, since one can mentally split by frontend/backend if desired to contain complexity, given the file structure).
* I think we probably want to make the setting being visible conditional on `server_web_public_streams_enabled`; that will prevent self-hosted servers from being offered an option that doesn't do anything. (I suppose in the future, we could have a way for self-hosted servers to submit themselves to Zulip Cloud for listing, which probably is a good idea, but until then it would just be confusing to have this setting).
* I'm thinking whether `advertise_in_communities_directory` is the right variable name for this, given that we have other filtering we'll do. Perhaps something like `want_advertise_in_communities_directory` would be most clear?

laurynmm commented:
@timabbott - Thanks for the review pass and those updates. I've squashed to one commit and updated the setting to be `want_advertise_in_communities_directory` throughout.

> - I think we probably want to make the setting being visible conditional on `server_web_public_streams_enabled`; that will prevent self-hosted servers from being offered an option that doesn't do anything. (I suppose in the future, we could have a way for self-hosted servers to submit themselves to Zulip Cloud for listing, which probably is a good idea, but until then it would just be confusing to have this setting).

I set the checkbox on the front end to have the same disable / enable functionality that the `realm_enable_spectator_access` checkbox has, which is based on that server setting you mentioned. Would you rather it be completely hidden when that server setting is on? If so, I can work on that next week. One thing I'll note is that I don't think there are any other checkboxes that change visibility (though there are a number of dropdowns that do).

timabbott commented:
I think copying the `realm_enable_spectator_access` behavior is reasonable for now; we can readily tweak in a follow-up as desired.

laurynmm commented:
> I think copying the `realm_enable_spectator_access` behavior is reasonable for now; we can readily tweak in a follow-up as desired.

That sound good to me. The screenshot below is when I set `WEB_PUBLIC_STREAMS_ENABLED` to False in the dev environment.

I did try to think of a synonym for "want to" for the setting name, but didn't come up with anything better or clearer. So the current commit has `want_advertise_in_communities_directory` as the setting name.

**Screenshot of setting UI when server has web public streams disabled**:
![Screenshot from 2022-05-02 19-23-58](https://user-images.githubusercontent.com/63245456/166295015-dd13f616-a1a6-461a-a78f-0bf2389fa628.png)

timabbott commented on static/templates/settings/organization_profile_admin.hbs:
```diff
@@ -25,6 +25,13 @@
+                  is_checked=realm_want_advertise_in_communities_directory
+                  is_disabled=disable_want_advertise_in_communities_directory
+                  label=admin_settings_label.realm_want_advertise_in_communities_directory
+                  help_link="help/communities-directory"}}
```
I think we want `/help/...` (note the leading slash); I also notice `git grep help_link=` shows we did this incorrectly with `help/public-access-option`.

In practice it doesn't matter with our current URL scheme, but it would be a latent bug that could be a problem in the future.

timabbott replied:
I added a commit addressing this via a linter extension in #21999.

laurynmm replied:
Thanks for catching that! And for adding the linter extension.

timabbott commented on templates/zerver/help/communities-directory.md:
```diff
@@ -0,0 +1,30 @@
+
+{settings_tab|organization-profile}
+
+1. Toggle **Advertise organization in the Zulip communities
```
This article should mention that the server needs to have enabled the public access option, as it is in Zulip Cloud, for this option to be enabled.

timabbott commented:
Posted a couple comments; otherwise this lgtm.

timabbott commented:
Merged via #21999, thanks @laurynmm!
